cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java

```diff
@@ -84,4 +108,70 @@ private static Configuration createConfiguration(Configuration hConf) {
     configuration.set(SparkContextConfig.HCONF_ATTR_EXECUTION_MODE, SparkContextConfig.YARN_EXECUTION_MODE);
     return configuration;
   }
+
+  /**
+   * Returns the {@link Resources} requirement for the workflow runnable due to spark driver resources requirement.
+   * Returns {@code null} if there is no spark program in the workflow.
+   */
+  @Nullable
+  private Resources findSparkDriverResources(Map<String, SparkSpecification> sparkSpecs, WorkflowSpecification spec) {
+    // Find the resource requirements from the workflow
+    // It is the largest memory and cores from all Spark program inside the workflow
+    Resources resources = new Resources();
+    boolean hasSpark = false;
+
+    // Search through all workflow nodes for spark program resource requirements.
+    Queue<WorkflowNode> nodes = new LinkedList<>(spec.getNodes());
+    while (!nodes.isEmpty()) {
+      WorkflowNode node = nodes.poll();
+      switch (node.getType()) {
+        case ACTION: {
+          ScheduleProgramInfo programInfo = ((WorkflowActionNode) node).getProgram();
+          if (programInfo.getProgramType() == SchedulableProgramType.SPARK) {
+            hasSpark = true;
+            // The sparkSpec shouldn't be null, otherwise the Workflow is not valid
+            Resources driverResources = sparkSpecs.get(programInfo.getProgramName()).getDriverResources();
+            if (driverResources != null) {
```

Comment:
It's only the spec shouldn't be null (meaning the `sparkSpecs.get(...)` part, that's why I just chain it instead of getting it back and check for null).